src/plonkish/backend/halo2.rs

```diff
@@ -187,193 +187,207 @@ impl<F: PrimeField + From<u64> + Hash> ChiquitoHalo2<F> {
-                            panic!("halo2 advice column not found {}", column.annotation)
-                        })
-                        .column;
-                    fixed_columns.insert(column.uuid(), halo2_column);
-                    meta.annotate_lookup_any_column(halo2_column, || column.annotation.clone());
-                }
-            }
-        }
-
-        self.advice_columns = advice_columns;
-        self.fixed_columns = fixed_columns;
-    }
+                Halo2Advice | Halo2Fixed => panic!(
+                    "Use src/plonkish/backend/halo2_legacy.rs to compile a circuit with imported Halo2 columns"
+                ),
+            });
 
-    fn configure_sub_circuit(&mut self, meta: &mut ConstraintSystem<F>) {
         if !self.plonkish_ir.exposed.is_empty() {
-            self.instance_column = Some(meta.instance_column());
+            meta.has_instance_column = true;
         }
 
-        if !self.plonkish_ir.polys.is_empty() {
-            meta.create_gate("main", |meta| {
-                let mut constraints: Vec<(&'static str, Expression<F>)> = Vec::new();
-
-                for poly in self.plonkish_ir.polys.iter() {
-                    let converted = self.convert_poly(meta, &poly.expr);
-                    let annotation = Box::leak(
-                        format!("{} => {:?}", poly.annotation, converted).into_boxed_str(),
-                    );
-                    constraints.push((annotation, converted));
-                }
-
-                constraints
-            });
-        }
+        self.plonkish_ir.polys.iter().for_each(|poly| {
+            meta.gates.push(GateMid {
+                name: "main".to_string(),
+                poly: meta.convert_poly(&poly.expr),
+            })
+        });
 
         for lookup in self.plonkish_ir.lookups.iter() {
             let annotation: &'static str = Box::leak(lookup.annotation.clone().into_boxed_str());
-            meta.lookup_any(annotation, |meta| {
-                let mut exprs = Vec::new();
-                for (src, dest) in lookup.exprs.iter() {
-                    exprs.push((self.convert_poly(meta, src), self.convert_poly(meta, dest)))
-                }
-
-                exprs
+            let mut exprs = Vec::new();
+            for (src, dest) in lookup.exprs.iter() {
+                exprs.push((meta.convert_poly(src), meta.convert_poly(dest)))
+            }
+            meta.lookups.push(lookup::ArgumentMid {
+                name: annotation.to_string(),
+                input_expressions: exprs.iter().map(|(src, _)| src.clone()).collect(),
+                table_expressions: exprs.iter().map(|(_, dest)| dest.clone()).collect(),
             });
         }
+
+        self.advice_columns = advice_columns;
+        self.fixed_columns = fixed_columns;
     }
+}
 
-    fn convert_poly(&self, meta: &mut VirtualCells<'_, F>, src: &PolyExpr<F>) -> Expression<F> {
+impl<F: PrimeField> ConstraintSystemBuilder<F> {
```

Comment:
@alxkzmn I think all the `impl` of ConstraintSystemBuilder should be after the `struct` defintion